File: cpp-impl/src/diagonal_traverse.cpp
#include <fmt/color.h>
#include <vector>
#include <algorithm>

using namespace std;

class Solution
{
public:
    static vector<int> findDiagonalOrder(vector<vector<int>>& mat)
    {
        int m = mat.size();
        int n = mat[0].size();
        vector<int> res, intermediate;
        int r = 0, c = 0;

        for (int d = 0; d < m + n - 1; ++d) {
            if (d < n) {
                r = 0;
                c = d;
            } else {
                r = d - n + 1;
                c = n - 1;
            }
            intermediate.clear();
            while (r < m && c > -1) {
                intermediate.push_back(mat[r][c]);
                ++r;
                --c;
            }
            if (intermediate.size() > 0 && d % 2 == 0) {
                reverse(intermediate.begin(), intermediate.end());
            }
            res.insert(res.end(), intermediate.begin(), intermediate.end());
        }
        return res;
    }
};

int testDiagonalTraverse() {
    vector<vector<int>> mat = {
            {1, 2, 3},
            {4, 5, 6},
            {7, 8, 9},
    };

    fmt::print(fg(fmt::color::green_yellow), "{}\n", fmt::join(Solution::findDiagonalOrder(mat), ", "));
    return 0;
}

File: rust-impl/src/diagonal_traverse.rs
struct Solution {}

impl Solution {
    pub fn find_diagonal_order(mat: Vec<Vec<i32>>) -> Vec<i32>{
        let mut res: Vec<i32> = Vec::new();
        let mut intermediate: Vec<i32> = Vec::new();
        let m: usize = mat.len();
        let n: usize = mat[0].len();
        let (mut r, mut c);
        for d in 0..(m + n - 1) {
            if d < n {
                r = 0;
                c = d;
            } else {
                r = d - n + 1;
                c = n - 1;
            }
            intermediate.clear();
            while r < m {
                intermediate.push(mat[r][c]);
                if c == 0 {
                    break
                }
                c -= 1;
                r += 1;
            }
            if intermediate.len() > 1 && d % 2 == 0 {
                intermediate.reverse();
            }
            res.extend(intermediate.iter());
        }
        return res;
    }
}

pub fn test_find_diagonal_order() {
    let mat: Vec<Vec<i32>> = vec![
        vec![1, 2, 3],
        vec![4, 5, 6],
        vec![7, 8, 9]
    ];
    let res: Vec<i32> = Solution::find_diagonal_order(mat);
    let mut str = String::new();
    for v in res {
        str.push_str(&format!("{}, ", v));
    }
    println!("{}", str.trim_end_matches(", "));
    return;
}